File: sw/resize.py
"""Software golden model for the resize stage.

Maps each output pixel (x, y) to a source coordinate (u, v) in the input image
using u = s_x * x, v = s_y * y, then performs bilinear interpolation from the
four nearest source pixels. Matches the formulation in plan.md.
"""

from __future__ import annotations

import numpy as np


def bilinear_sample(img: np.ndarray, u: float, v: float) -> np.ndarray:
    H, W = img.shape[:2]

    u = np.clip(u, 0.0, W - 1.0)
    v = np.clip(v, 0.0, H - 1.0)

    u0 = int(np.floor(u))
    v0 = int(np.floor(v))
    u1 = min(u0 + 1, W - 1)
    v1 = min(v0 + 1, H - 1)

    du = np.float32(u - u0)
    dv = np.float32(v - v0)

    w00 = (np.float32(1.0) - du) * (np.float32(1.0) - dv)
    w10 = du * (np.float32(1.0) - dv)
    w01 = (np.float32(1.0) - du) * dv
    w11 = du * dv

    p00 = img[v0, u0].astype(np.float32)
    p10 = img[v0, u1].astype(np.float32)
    p01 = img[v1, u0].astype(np.float32)
    p11 = img[v1, u1].astype(np.float32)

    return w00 * p00 + w10 * p10 + w01 * p01 + w11 * p11


def resize(img_in: np.ndarray, out_h: int, out_w: int) -> np.ndarray:
    """Resize img_in to (out_h, out_w) with bilinear interpolation.

    Follows the u = s_x * x, v = s_y * y formulation from plan.md (no
    half-pixel centering) so the hardware coordinate generator can use the
    same arithmetic.
    """
    in_h, in_w = img_in.shape[:2]
    scale_x = np.float32(in_w / out_w)
    scale_y = np.float32(in_h / out_h)

    out_shape = (out_h, out_w) + img_in.shape[2:]
    out = np.zeros(out_shape, dtype=np.float32)

    for y in range(out_h):
        v = scale_y * np.float32(y)
        for x in range(out_w):
            u = scale_x * np.float32(x)
            out[y, x] = bilinear_sample(img_in, u, v)

    return np.clip(out, 0, 255).astype(img_in.dtype)


def resize_vectorized(img_in: np.ndarray, out_h: int, out_w: int) -> np.ndarray:
    """Vectorized equivalent of resize() — same math, faster for large images."""
    in_h, in_w = img_in.shape[:2]
    scale_x = np.float32(in_w / out_w)
    scale_y = np.float32(in_h / out_h)

    xs = np.clip(np.arange(out_w, dtype=np.float32) * scale_x, 0.0, np.float32(in_w - 1))
    ys = np.clip(np.arange(out_h, dtype=np.float32) * scale_y, 0.0, np.float32(in_h - 1))

    u0 = np.floor(xs).astype(np.int32)
    v0 = np.floor(ys).astype(np.int32)
    u1 = np.minimum(u0 + 1, in_w - 1)
    v1 = np.minimum(v0 + 1, in_h - 1)

    du = (xs - u0.astype(np.float32)).reshape(1, -1)
    dv = (ys - v0.astype(np.float32)).reshape(-1, 1)

    img_f = img_in.astype(np.float32)
    p00 = img_f[np.ix_(v0, u0)]
    p10 = img_f[np.ix_(v0, u1)]
    p01 = img_f[np.ix_(v1, u0)]
    p11 = img_f[np.ix_(v1, u1)]

    if img_in.ndim == 3:
        du = du[..., None]
        dv = dv[..., None]

    one = np.float32(1.0)
    w00 = (one - du) * (one - dv)
    w10 = du * (one - dv)
    w01 = (one - du) * dv
    w11 = du * dv

    out = w00 * p00 + w10 * p10 + w01 * p01 + w11 * p11
    return np.clip(out, 0, 255).astype(img_in.dtype)

File: hls/resize.cpp
#include "resize.h"

#include <hls_math.h>

// Baseline streaming resize with bilinear interpolation.
//
// Coordinate model (matches sw/resize.py):
//   u = scale_x * x   (no half-pixel centering)
//   v = scale_y * y
// then sample img_in at (u, v) with bilinear interpolation. Coordinates are
// clamped to [0, W-1] × [0, H-1] so out-of-range lookups replicate borders.
//
// Schedule: we keep a 2-row ring buffer of input rows. Because v is monotonic
// non-decreasing in the output row index y, any output row y only needs input
// rows v0(y) and v0(y)+1. After reading input row r, we emit every output row
// y whose v0(y)+1 == r (or for upscale, every y whose pair is [r-1, r] that
// we haven't emitted yet). This makes each input row readable exactly once
// from the AXI-Stream and each output pixel writable exactly once — the fast
// path the plan calls for.

static pixel_t bilinear(
    pixel_t p00, pixel_t p10, pixel_t p01, pixel_t p11,
    float du, float dv)
{
#pragma HLS INLINE
    float one_du = 1.0f - du;
    float one_dv = 1.0f - dv;
    float w00 = one_du * one_dv;
    float w10 = du     * one_dv;
    float w01 = one_du * dv;
    float w11 = du     * dv;

    float acc = w00 * (float)p00
              + w10 * (float)p10
              + w01 * (float)p01
              + w11 * (float)p11;

    if (acc < 0.0f)   acc = 0.0f;
    if (acc > 255.0f) acc = 255.0f;
    return (pixel_t)(ap_uint<8>)(int)acc;  // truncate-to-uint8 like Python's astype
}

void resize_kernel(
    hls::stream<pixel_t> &in_stream,
    hls::stream<pixel_t> &out_stream,
    dim_t in_w,
    dim_t in_h,
    dim_t out_w,
    dim_t out_h,
    float scale_x,
    float scale_y)
{
#pragma HLS INTERFACE axis port=in_stream
#pragma HLS INTERFACE axis port=out_stream
#pragma HLS INTERFACE s_axilite port=in_w     bundle=ctrl
#pragma HLS INTERFACE s_axilite port=in_h     bundle=ctrl
#pragma HLS INTERFACE s_axilite port=out_w    bundle=ctrl
#pragma HLS INTERFACE s_axilite port=out_h    bundle=ctrl
#pragma HLS INTERFACE s_axilite port=scale_x  bundle=ctrl
#pragma HLS INTERFACE s_axilite port=scale_y  bundle=ctrl
#pragma HLS INTERFACE s_axilite port=return   bundle=ctrl

    // 2-row ring buffer, partitioned so both rows can be read in the same
    // cycle to evaluate a 2x2 neighborhood at II=1.
    static pixel_t line_buf[2][RESIZE_MAX_IN_W];
#pragma HLS ARRAY_PARTITION variable=line_buf dim=1 complete
#pragma HLS BIND_STORAGE variable=line_buf type=ram_t2p impl=bram

    dim_t next_y = 0;

    INPUT_ROW: for (dim_t r = 0; r < in_h; r++) {
#pragma HLS LOOP_TRIPCOUNT min=1 max=RESIZE_MAX_IN_H

        // Load input row r into the ring slot r&1.
        LOAD_ROW: for (dim_t xi = 0; xi < in_w; xi++) {
#pragma HLS LOOP_TRIPCOUNT min=1 max=RESIZE_MAX_IN_W
#pragma HLS PIPELINE II=1
            line_buf[r & 1][xi] = in_stream.read();
        }

        // Emit every output row whose (v0, v1) pair is now in the buffer.
        // After loading row r, rows (r-1, r) and (r, r) (bottom-edge clamp)
        // are available.
        EMIT: while (next_y < out_h) {
#pragma HLS LOOP_TRIPCOUNT min=0 max=RESIZE_MAX_OUT_H

            float v_f = scale_y * (float)(int)next_y;
            // clamp v_f to [0, in_h - 1]
            float v_max = (float)((int)in_h - 1);
            if (v_f < 0.0f)    v_f = 0.0f;
            if (v_f > v_max)   v_f = v_max;

            int v0i = (int)v_f;  // floor for non-negative
            int v1i = v0i + 1;
            if (v1i > (int)in_h - 1) v1i = (int)in_h - 1;

            // Need rows v0i and v1i both <= r before emitting.
            if (v1i > (int)r) break;

            float dv = v_f - (float)v0i;
            ap_uint<1> row0_slot = v0i & 1;
            ap_uint<1> row1_slot = v1i & 1;

            EMIT_PIXEL: for (dim_t xo = 0; xo < out_w; xo++) {
#pragma HLS LOOP_TRIPCOUNT min=1 max=RESIZE_MAX_OUT_W
#pragma HLS PIPELINE II=1
                float u_f = scale_x * (float)(int)xo;
                float u_max = (float)((int)in_w - 1);
                if (u_f < 0.0f)  u_f = 0.0f;
                if (u_f > u_max) u_f = u_max;

                int u0i = (int)u_f;
                int u1i = u0i + 1;
                if (u1i > (int)in_w - 1) u1i = (int)in_w - 1;

                float du = u_f - (float)u0i;

                pixel_t p00 = line_buf[row0_slot][u0i];
                pixel_t p10 = line_buf[row0_slot][u1i];
                pixel_t p01 = line_buf[row1_slot][u0i];
                pixel_t p11 = line_buf[row1_slot][u1i];

                out_stream.write(bilinear(p00, p10, p01, p11, du, dv));
            }

            next_y++;
        }
    }
}
